README.md
# clojure-synth

An app designed to output monophonic sound waves (sine, square, sawtooth).

This app generates basic sound waves to generate monophonic tones. To play a tone or a melody, you have to supply notes; they are represented as vectors with three attributes - pitch, volume and duration. These attributes are based on something you might find in a sheet music programs. 

Pitch is is represented as as string containing a note and it's octave (eg A4). App is based around A4 pitch set to 440Hz. All the other notes are then calculated on the spot, meaning you can change the value of A4 and all other notes will change during execution, based on equal temperament system. Pitches range from C0 to B8, any other pitches are unsupported and will throw an error.

Volume can be in the range from 0 to a 100. The lowest value - 0 - can be used as a pause.

Duration consist of 7 posible numbers to choose from: 1, 2, 4, 8, 16, 32, 64. These numbers represent the denominator of a fraction that represents note's duration. Quarter notes are the base of all other duration. They are represented as 4. When a certain bpm is supplied, app calculates the duration of a quarter note in milliseconds. After that all other durations are calculated by their ratio to the quarter note.

## Quick Start

If you want to test this app, you can play the default tune easily! Type in lein run and you will be prompted to press enter. After that, a short audio sample will ring out. Alternatively, follow the instructions to play your melody.

## Usage

Playing through command line:

1. Type in lein run in your CLI.

2. Test out predefined melody by typing in test or press enter to start making your melody.

3. Enter beats per minute; value needs to be in range from 30 to 300.

4. Choose oscillators by entering the corresponding number (1 - sine, 2 - square, 3 - sawtooth).

5. Start inputing notes: pitch values need to be between C0 and B8, volume needs to be between 0 and 100 (0 can be used as a pause) and duration must be one of the following values: 1, 2, 4, 8, 16, 32, 64 - numbers represent the denominator of a fraction, eg. 4 is 1/4 or quarter note.

6. When you are done inputing your melody type in done. Melody will play. 

7. After that you can continue using the app from the start or exit anytime by typing in exit.

Playing through code:

1. Define a custom format using audio-formats/custom-format function and pass sample rate, number of bits and number of channels as arguments
    OR 
    Use default format which is already defined as "af" variable.

2. Make your own melody by defining a vector containing note vectors. A note vector consists of pitch (C0 - B8), volume (0 - 100) and duration (1, 2, 4, 8, 16, 32, 64 - numbers represent the denominator of a fraction, eg. 4 is 1/4 or quarter note). You can use any pitch combined with 0 volume to get a pause. Eg. ["B4" 100 8]
    OR
    Use predefined melodies. They are stored in variables (rondo-alla-turca or samo-mi-kazi).

3. Choose oscillator function to be used for playback. Currently available oscillators are sine, square and sawtooth.

4. Play the melody! You can play your or predefined melody by calling the play-melody function and passing in audio format, oscillator, melody and beats per minute. Eg (play-melody af sine rondo-alla-turca 260)

# Future updates

This app is easy to expand with many features. Some of those features are:
- More oscillators
- Playing the melody indefinitely or reacting to a key press on keyboard (therefore, computer keyboard can be used as an actual muscal keyboard)
- Dotted durations
- Erasing notes in CLI melody input
- Case insensitive pitch

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-12-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-12-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/clojure-synth/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/clojure-synth/compare/0.1.0...0.1.1


src/clojure_synth/core.clj
(ns clojure-synth.core
  (:require [clojure-synth.audio-formats :as audio-formats]
            [clojure-synth.oscillators :as oscillators]
            [clojure-synth.player :as player]
            [clojure-synth.sample-melodies :as sample-melodies]
            [clojure-synth.duration-helpers :as duration-helpers]
            [clojure-synth.frequency-helpers :as frequency-helpers]
            [clojure-synth.volume-helpers :as volume-helpers]))

(def af audio-formats/default-format)
(def sine oscillators/sine)
(def saw oscillators/saw)
(def square oscillators/square)
(def samo-mi-kazi sample-melodies/samo-mi-kazi)
(def rondo-alla-turca sample-melodies/rondo-alla-turca)

(def play-melody player/play-melody)
;; (play-melody audio-format oscillator melody bpm)
;; (play-melody af sine rondo-alla-turca 260)
;; (play-melody af sine samo-mi-kazi 165)

(defn -main
  []
  (println "Clojure-Synth has started!")
  (while true 
    (println "Press enter to start using the app or type in 'test' to play the default tune! You can exit anytime by typing exit.")
    (let [in (read-line)]
      (case in
        "test" (play-melody af sine samo-mi-kazi 165)
        "exit" (java.lang.System/exit 0)
        ""
        (with-local-vars [bpm 120
                          melody []
                          oscillator sine
                          oscillator-name "sine"
                          melody-set false
                          bpm-valid false
                          osc-valid false]
          (println "Please enter bpm (30 - 300):")
          (while (not (var-get bpm-valid))
            (let [b (read-line)]
              (when (= b "exit")
                (java.lang.System/exit 0))
              (if (and (every? #(Character/isDigit %) b) (duration-helpers/bpm-validator (Integer/parseInt b)))
                (do (var-set bpm (Integer/parseInt b))
                    (var-set bpm-valid true))
                (println "Invalid bpm value, try again:"))))
          (println "Please choose an oscillator (1 - sine, 2 - square, 3 - sawtooth):")
          (while (not (var-get osc-valid))
            (let [osc (read-line)]
              (case osc
                "1" (do (var-set oscillator sine)
                        (var-set osc-valid true)
                        (var-set oscillator-name "sine"))
                "2" (do (var-set oscillator square)
                        (var-set osc-valid true)
                        (var-set oscillator-name "square"))
                "3" (do (var-set oscillator saw)
                        (var-set osc-valid true)
                        (var-set oscillator-name "sawtooth"))
                "exit" (java.lang.System/exit 0)
                (println "Invalid oscillator, try again:"))))
          (while (not (var-get melody-set))
            (with-local-vars [note []
                              pitch-valid false
                              volume-valid false
                              duration-valid false]
              (println "Please enter pitch (C0 - B8) or type in done to finish:")
              (while (not (var-get pitch-valid))
                (let [n (read-line)]
                  (when (= n "exit")
                    (java.lang.System/exit 0))
                  (if (= n "done")
                    (do (var-set melody-set true)
                        (var-set pitch-valid true))
                    (do (if (frequency-helpers/pitch-validator n)
                          (do (var-set note (conj (var-get note) n))
                              (var-set pitch-valid true)
                              (println "Please enter volume (0 - 100):")
                              (while (not (var-get volume-valid))
                                (let [v (read-line)]
                                  (when (= v "exit")
                                    (java.lang.System/exit 0))
                                  (if (and (every? #(Character/isDigit %) v) (volume-helpers/volume-validator (Integer/parseInt v)))
                                    (do (var-set note (conj (var-get note) (Integer/parseInt v)))
                                        (var-set volume-valid true))
                                    (println "Invalid volume, try again:"))))
                              (println "Please enter duration (1, 2, 4, 8, 16, 32, 64):")
                              (while (not (var-get duration-valid))
                                (let [d (read-line)]
                                  (when (= d "exit")
                                    (java.lang.System/exit 0))
                                  (if (and (every? #(Character/isDigit %) d) (duration-helpers/duration-validator (Integer/parseInt d)))
                                    (do (var-set note (conj (var-get note) (Integer/parseInt d)))
                                        (var-set duration-valid true))
                                    (println "Invalid duration, try again:"))))
                              (var-set melody (conj (var-get melody) (var-get note)))
                              (println (var-get melody)))
                          (println "Invalid pitch, try again:"))))))))
          (println (str "Playing melody at " (var-get bpm) " bpms, using " (var-get oscillator-name) " oscillator."))
          (play-melody af (var-get oscillator) (var-get melody) (var-get bpm)))
        (println "Invalid command, try again:")))))